File: backend/src/modules/billable-metrics/dto/update-billable-metric.dto.ts
import {
  IsString,
  IsOptional,
  IsBoolean,
  IsArray,
  ValidateNested,
  MaxLength,
} from 'class-validator';
import { ApiProperty } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import { CreateBillableMetricFilterDto } from './create-billable-metric.dto';

export class UpdateBillableMetricDto {
  @ApiProperty({ required: false, example: 'API Requests' })
  @IsString()
  @IsOptional()
  @MaxLength(100)
  name?: string;

  @ApiProperty({ required: false, example: 'Number of API requests' })
  @IsString()
  @IsOptional()
  @MaxLength(500)
  description?: string;

  @ApiProperty({ required: false, example: 'tokens' })
  @IsString()
  @IsOptional()
  fieldName?: string;

  @ApiProperty({ required: false })
  @IsBoolean()
  @IsOptional()
  recurring?: boolean;

  @ApiProperty({ required: false, type: [CreateBillableMetricFilterDto] })
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => CreateBillableMetricFilterDto)
  @IsOptional()
  filters?: CreateBillableMetricFilterDto[];
}

File: sdks/csharp/src/NovaBilling/BillableMetrics/Requests/UpdateBillableMetricDto.cs
using System.Text.Json.Serialization;
using NovaBilling.Core;

namespace NovaBilling;

[Serializable]
public record UpdateBillableMetricDto
{
    /// <summary>
    /// Billable Metric ID
    /// </summary>
    [JsonIgnore]
    public required string Id { get; set; }

    [JsonPropertyName("name")]
    public string? Name { get; set; }

    [JsonPropertyName("description")]
    public string? Description { get; set; }

    [JsonPropertyName("fieldName")]
    public string? FieldName { get; set; }

    [JsonPropertyName("recurring")]
    public bool? Recurring { get; set; }

    [JsonPropertyName("filters")]
    public IEnumerable<CreateBillableMetricFilterDto>? Filters { get; set; }

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}
